src/channel/httpBeaconChannel.ts

```diff
@@ -0,0 +1,115 @@
+import {ChannelListener, DuplexChannel} from './channel';
+import {Envelope} from './guaranteedChannel';
+import {Logger, NullLogger} from '../logging';
+import {CidAssigner} from '../cid';
+import {formatMessage} from '../error';
+
+export type Configuration = {
+    appId: string,
+    endpointUrl: string,
+    cidAssigner: CidAssigner,
+    logger?: Logger,
+};
+
+type ApiProblem = {
+    type: string,
+    title: string,
+    status: number,
+    detail: string,
+};
+
+export class HttpBeaconChannel implements DuplexChannel<string, Envelope<string, string>> {
+    private static readonly NON_RETRYABLE_STATUSES: ReadonlySet<number> = new Set([
+        403, // API usage limit exceeded
+        401, // Invalid token
+    ]);
+
+    private readonly configuration: Omit<Configuration, 'logger'>;
+
+    private readonly logger: Logger;
+
+    private readonly listeners: Array<ChannelListener<string>> = [];
+
+    private closed = false;
+
+    public constructor({logger = new NullLogger(), ...configuration}: Configuration) {
+        this.configuration = configuration;
+        this.logger = logger;
+    }
+
+    public async publish({id: receiptId, message}: Envelope<string, string>): Promise<void> {
+        if (this.closed) {
+            return Promise.reject(new Error('Channel is closed'));
+        }
+
+        const {token, timestamp, context, payload} = JSON.parse(message);
+        const {endpointUrl, appId, cidAssigner} = this.configuration;
+
+        return fetch(endpointUrl, {
+            method: 'POST',
+            headers: {
+                'X-Application-Id': appId,
+                'X-Client-Id': await cidAssigner.assignCid(),
+                ...(token !== undefined ? {'X-Token': token} : {}),
+            },
+            body: JSON.stringify({
+                context: context,
+                payload: payload,
+                originalTime: timestamp,
+                departureTime: Date.now(),
+            }),
+        }).then(async response => {
+            if (response.ok) {
+                this.notify(receiptId);
+
+                return;
+            }
+
+            const problem: ApiProblem = await response.json().catch(
+                () => ({
+                    type: 'https://croct.help/api/event-tracker#unexpected-error',
+                    title: response.statusText,
+                    status: response.status,
+                }),
+            );
+
+            if (HttpBeaconChannel.NON_RETRYABLE_STATUSES.has(problem.status)) {
```

Comment:
Why get just the first violation now?

Reply:
I'm not sure if I understood your question.

Reply:
Before we received an [array of violations](https://github.com/croct-tech/sdk-js/pull/410/files/5d3bde37aea140bd5f97ca52f5b0f7ff29aeb52a#diff-d03778d1a54ef4507f22532992473ee8001d3865a5b94aa84d49743a83d9e79cL139) and logged all of them, now we get just the first one

Reply:
I don't think we were doing that before.

Reply:
In theory, this should never happen.